### ISSUE TITLE
Can't create event loop ondemand

### ISSUE THREAD

**Issue by emilk** [rust-windowing/winit#2431]
I want to be able to open a window, use `event_loop.run_return` to process some events on it, then close it, and then later open another window. When I try this on my Mac, I find that the second time I run `run_return` I get `Event::LoopDestroyed` almost immediately, and the new window is destroyed. This is the code I'm running:

```rust
fn main() {
    use winit::{
        event::{Event, WindowEvent},
        event_loop::EventLoop,
        platform::run_return::EventLoopExtRunReturn,
        window::WindowBuilder,
    };

    fn show_window(event_loop: &mut EventLoop<()>, window_name: &str) {
        eprintln!("Creating window {:?}", window_name);

        let window = WindowBuilder::new()
            .with_title(window_name)
            .build(event_loop)
            .unwrap();

        event_loop.run_return(|event, _, control_flow| {
            control_flow.set_wait();

            match event {
                Event::LoopDestroyed => {
                    eprintln!("Event::LoopDestroyed.");
                }

                // Platform-dependent event handlers to workaround a winit bug
                // See: https://github.com/rust-windowing/winit/issues/987
                // See: https://github.com/rust-windowing/winit/issues/1619
                Event::RedrawEventsCleared if cfg!(windows) => {
                    println!("rendering");
                }
                Event::RedrawRequested(_) if !cfg!(windows) => {
                    println!("rendering");
                }

                Event::WindowEvent { window_id, .. } if window_id != window.id() => {
                    // This can happen if we close a window, and then reopen a new one,
                    // or if we have multiple windows open.
                    eprintln!("Ignoring event to old window");
                }
                Event::WindowEvent {
                    event: WindowEvent::CloseRequested,
                    ..
                } => {
                    eprintln!("WindowEvent::CloseRequested");
                    control_flow.set_exit();
                }
                Event::WindowEvent {
                    event: WindowEvent::Destroyed,
                    ..
                } => {
                    eprintln!("WindowEvent::Destroyed");
                    control_flow.set_exit();
                }
                _ => (),
            }
        });

        eprintln!("-----------------");
    }

    let mut event_loop = EventLoop::new();
    show_window(&mut event_loop, "First Window");
    show_window(&mut event_loop, "Second Window");
}
```

And this is the output:

```
Creating window "First Window"
rendering
rendering
WindowEvent::CloseRequested
Event::LoopDestroyed.
-----------------
Creating window "Second Window"
rendering
rendering
Event::LoopDestroyed.
-----------------
```

The `WindowEvent::CloseRequested` is me closing the window, but the second window also closes without any reason.

According to @zu1k, this is also a problem on Windows, but there it can be fixed by adding the following code after the first `run_return`:

```rust
    event_loop.run_return(|_, _, control_flow| {
        control_flow.set_exit();
    });
```

Related issue: https://github.com/emilk/egui/pull/1919

---
One very bad work-around is to calli `run_return` multiple times, like so:

``` rust
        let mut quit = false;
        while !quit {
            event_loop.run_return(|event, _, control_flow| {
                // Only set `quit = true;` when we really want to close the window.
            });
        }
```

This still spams `Event::LoopDestroyed` over and over again, but the window stays up. However, it uses up ALL the CPU of one core, so this is NOT a viable solution.

**Comment by kchibisov** [comment]
Not sure how I've missed that issue.

You're might not be familiar with such concept, but event loop is a connection to the display server in the first place, dropping it will drop all the windows present, it's more of an exception on macOS, it's not possible at all on X11/Wayland.

I think that `LoopDestroyed` is confusing event, the loop is not destroyed with `run_return` at all, it's alive, it's just not polling the events. 

>One very bad work-around is to calli run_return multiple times, like so:

It seems like what you want is to run winit's event loop along side some other event loop and not handle everything inside the `run``run_return` closure. Such use case is sort of common when it comes to polling, from what I can tell.

There's an example how to do `block_on` https://github.com/notgull/winit-block-on. In your case I'd assume you could simply set the value inside the `run_return` and `Wait` in it initially.

>This still spams Event::LoopDestroyed over and over again, but the window stays up. However, it uses up ALL the CPU of one core, so this is NOT a viable solution.

It'll be a solution if you `Wait` and exit on some condition you set from the `run_return`.

The issue right now is that you're not notified for winit's event loop readiness.


So, to sum up, what you want is not possibly theoretically, but it's likely not what you actually need anyway. There's no solution on putting winit's loop inside your `epoll`, and I doubt there will be a good way to do this... The alternative solution to register inside winit's epoll, which is more of a realistic, once the macOS/Windows will be sorted.

Maybe approach with `async` and `block_on` will work for you though.

**Comment by kchibisov** [comment]
The issue title is also misleading.

**Comment by rib** [comment]
> > One very bad work-around is to calli run_return multiple times, like so:
> 
> It seems like what you want is to run winit's event loop along side some other event loop and not handle everything inside the `` run``run_return `` closure. Such use case is sort of common when it comes to polling, from what I can tell.

I think this is describing what the work-around is doing, not what the original intent seems to be.

E.g. see here for added context for this issue: https://github.com/emilk/egui/pull/1889 (based on other issues that @emilk has linked here)

I think for the use case behind this issue @emilk would have been happy for all the event handling to be done inside `run_return`.

@kchibisov  I think you discovered this issue after I linked it to this issue https://github.com/rust-windowing/winit/issues/2706#issuecomment-1455945744 and I did that in part because this was an example of a use case that _didn't_ want to move handling of events outside of winit. This issue looks like its about being able to run a temporary (but self-contained) event loop.

I think the use case here is being able to have some application (hypothetically could be a command line app, not based on a mainloop) that shows a temporary pop-up UI, e.g. to show stats.

I think for this use case the `run_return` was expected to be self-contained and behave more-or-less like `run` except for the fact that the event loop could exit after the temporary UI has been closed, and potentially run again in the future.

For this use case it's potentially fine that it will completely close any display connections etc when the loop exits because the loop will only exit after the temporary UI has been closed.

So I don't think there was any expectation here that `run_return` was going to return while leaving windows / UI intact, it's just the "bad work-around" that relies on this.

**Comment by rib** [comment]
I think maybe it could be fair to re-open this issue while it looks like the reason for closing it was more based on how the workaround works, not so much about the original issue.

The response above looks like it's more applicable to https://github.com/rust-windowing/winit/issues/2706 than here, since the code in #2706 really is explicitly trying to embed Winit event polling into an outer loop.

**Comment by rib** [comment]
> I think I've clearly stated why it doesn't applicable? Dropping event loop will drop connection to display server, dropping all the windows along the way. I'm pretty sure you're aware what happens when you close connection to x11/wayland server? I'm not entirely sure wrt X11, but with Wayland everything will die.

yeah, but for this use case I think that's ok. It's sort of like you're responding to the problems with #2706 here.

I think here the event loop is only intended to exit after the windows for the temporary UI have been closed, so there won't be any windows to preserve and it's fine that the display connection is dropped.

**Comment by kchibisov** [comment]
Oh, I seems like misread the issue example slightly, indeed looks like a macOS issue. You should be able to open window, do things with it, drop, and open again latter on with different `run_return` invocation.

**Comment by kchibisov** [comment]
@emilk could you try to drop window from inside the event loop and see if it changes anything or if the latest winit has this issue in the first place? I think on macOS and some other platforms you need event loop running to things work properly.

**Comment by emilk** [comment]
Trying the original code on latest `master` (b18295a1ce1efbf1d0ab6551b5df1d7e6c511d50) I get similar results:

```
Creating window "First Window"
rendering
rendering
WindowEvent::CloseRequested
Event::LoopDestroyed.
-----------------
Creating window "Second Window"
rendering
Ignoring event to old window
rendering
Event::LoopDestroyed.
-----------------
```

I've tried dropping the `Window` from inside the event loop like so:

```rust
fn main() {
    use winit::{
        event::{Event, WindowEvent},
        event_loop::EventLoop,
        platform::run_return::EventLoopExtRunReturn,
        window::WindowBuilder,
    };

    fn show_window(event_loop: &mut EventLoop<()>, window_name: &str) {
        eprintln!("Creating window {:?}", window_name);

        let window = WindowBuilder::new()
            .with_title(window_name)
            .build(event_loop)
            .unwrap();
        let our_window_id = window.id();
        let mut window = Some(window);

        event_loop.run_return(|event, _, control_flow| {
            control_flow.set_wait();

            match event {
                Event::LoopDestroyed => {
                    eprintln!("Event::LoopDestroyed.");
                    window = None;
                }

                // Platform-dependent event handlers to workaround a winit bug
                // See: https://github.com/rust-windowing/winit/issues/987
                // See: https://github.com/rust-windowing/winit/issues/1619
                Event::RedrawEventsCleared if cfg!(windows) => {
                    println!("rendering");
                }
                Event::RedrawRequested(_) if !cfg!(windows) => {
                    println!("rendering");
                }

                Event::WindowEvent { window_id, .. } if window_id != our_window_id => {
                    // This can happen if we close a window, and then reopen a new one,
                    // or if we have multiple windows open.
                    eprintln!("Ignoring event to old window");
                }
                Event::WindowEvent {
                    event: WindowEvent::CloseRequested,
                    ..
                } => {
                    eprintln!("WindowEvent::CloseRequested");
                    window = None;
                    control_flow.set_exit();
                }
                Event::WindowEvent {
                    event: WindowEvent::Destroyed,
                    ..
                } => {
                    eprintln!("WindowEvent::Destroyed");
                    window = None;
                    control_flow.set_exit();
                }
                _ => (),
            }
        });

        eprintln!("-----------------");
    }

    let mut event_loop = EventLoop::new();
    show_window(&mut event_loop, "First Window");
    show_window(&mut event_loop, "Second Window");
}
```

But the result is still the same - I only get one window, and the output is still

```
Creating window "First Window"
rendering
rendering
WindowEvent::CloseRequested
Event::LoopDestroyed.
-----------------
Creating window "Second Window"
rendering
Ignoring event to old window
rendering
Event::LoopDestroyed.
-----------------
```

**Comment by kchibisov** [comment]
What if you create event loop each time for a new window, will it work?

**Comment by emilk** [comment]
With

```diff
--- a/src/event_loop.rs
+++ b/src/event_loop.rs
@@ -112,9 +112,9 @@ impl<T> EventLoopBuilder<T> {
     #[inline]
     pub fn build(&mut self) -> EventLoop<T> {
         static EVENT_LOOP_CREATED: OnceCell<()> = OnceCell::new();
-        if EVENT_LOOP_CREATED.set(()).is_err() {
-            panic!("Creating EventLoop multiple times is not supported.");
-        }
+        // if EVENT_LOOP_CREATED.set(()).is_err() {
+        //     panic!("Creating EventLoop multiple times is not supported.");
+        // }
         // Certain platforms accept a mutable reference in their API.
         #[allow(clippy::unnecessary_mut_passed)]
         EventLoop {
```

and recreating the event-loop I get:

```
❯ RUST_BACKTRACE=1 RUST_LOG=debug cargo r --example emilk
warning: static `EVENT_LOOP_CREATED` is never used
   --> src/event_loop.rs:114:16
    |
114 |         static EVENT_LOOP_CREATED: OnceCell<()> = OnceCell::new();
    |                ^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(dead_code)]` on by default

warning: `winit` (lib) generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.04s
     Running `target/debug/examples/emilk`
Creating window "First Window"
TRACE [winit::platform_impl::platform::window] Triggered `WinitWindow::new`
TRACE [winit::platform_impl::platform::view] Triggered `validAttributesForMarkedText`
TRACE [winit::platform_impl::platform::view] Completed `validAttributesForMarkedText`
TRACE [winit::platform_impl::platform::view] Triggered `validAttributesForMarkedText`
TRACE [winit::platform_impl::platform::view] Completed `validAttributesForMarkedText`
TRACE [winit::platform_impl::platform::view] Triggered `validAttributesForMarkedText`
TRACE [winit::platform_impl::platform::view] Completed `validAttributesForMarkedText`
TRACE [winit::platform_impl::platform::view] Triggered `frameDidChange:`
TRACE [winit::platform_impl::platform::view] Completed `frameDidChange:`
TRACE [winit::platform_impl::platform::view] Triggered `viewDidMoveToWindow`
TRACE [winit::platform_impl::platform::view] Completed `viewDidMoveToWindow`
TRACE [winit::platform_impl::platform::window] Locked shared state in `WinitWindow::new`
TRACE [winit::platform_impl::platform::window] Unlocked shared state in `WinitWindow::new`
TRACE [winit::platform_impl::platform::window] Locked shared state in `set_fullscreen`
TRACE [winit::platform_impl::platform::window] Unlocked shared state in `set_fullscreen`
TRACE [winit::platform_impl::platform::window] Triggered `canBecomeMainWindow`
TRACE [winit::platform_impl::platform::window] Completed `canBecomeMainWindow`
TRACE [winit::platform_impl::platform::window] Triggered `canBecomeKeyWindow`
TRACE [winit::platform_impl::platform::window] Completed `canBecomeKeyWindow`
TRACE [winit::platform_impl::platform::view] Triggered `acceptsFirstResponder`
TRACE [winit::platform_impl::platform::view] Completed `acceptsFirstResponder`
TRACE [winit::platform_impl::platform::view] Triggered `acceptsFirstResponder`
TRACE [winit::platform_impl::platform::view] Completed `acceptsFirstResponder`
TRACE [winit::platform_impl::platform::window] Triggered `canBecomeKeyWindow`
TRACE [winit::platform_impl::platform::window] Completed `canBecomeKeyWindow`
TRACE [winit::platform_impl::platform::window] Triggered `canBecomeKeyWindow`
TRACE [winit::platform_impl::platform::window] Completed `canBecomeKeyWindow`
TRACE [winit::platform_impl::platform::window] Completed `WinitWindow::new`
TRACE [winit::platform_impl::platform::view] Triggered `drawRect:`
rendering
TRACE [winit::platform_impl::platform::view] Completed `drawRect:`
TRACE [winit::platform_impl::platform::app_delegate] Triggered `applicationDidFinishLaunching:`
TRACE [winit::platform_impl::platform::app_state] Activating visible window
TRACE [winit::platform_impl::platform::window] Triggered `canBecomeMainWindow`
TRACE [winit::platform_impl::platform::window] Completed `canBecomeMainWindow`
TRACE [winit::platform_impl::platform::window] Triggered `canBecomeKeyWindow`
TRACE [winit::platform_impl::platform::window] Completed `canBecomeKeyWindow`
TRACE [winit::platform_impl::platform::window] Triggered `canBecomeKeyWindow`
TRACE [winit::platform_impl::platform::window] Completed `canBecomeKeyWindow`
TRACE [winit::platform_impl::platform::app_delegate] Completed `applicationDidFinishLaunching:`
rendering
TRACE [winit::platform_impl::platform::window] Triggered `canBecomeKeyWindow`
TRACE [winit::platform_impl::platform::window] Completed `canBecomeKeyWindow`
TRACE [winit::platform_impl::platform::window_delegate] Triggered `windowDidBecomeKey:`
TRACE [winit::platform_impl::platform::window_delegate] Completed `windowDidBecomeKey:`
TRACE [winit::platform_impl::platform::view] Triggered `resetCursorRects`
TRACE [winit::platform_impl::platform::view] Completed `resetCursorRects`
TRACE [winit::platform_impl::platform::window_delegate] Triggered `windowDidChangeOcclusionState:`
TRACE [winit::platform_impl::platform::window_delegate] Completed `windowDidChangeOcclusionState:`
TRACE [winit::platform_impl::platform::window_delegate] Triggered `windowDidChangeScreen:`
TRACE [winit::platform_impl::platform::window] Locked shared state in `window_did_change_screen`
TRACE [winit::platform_impl::platform::window] Unlocked shared state in `window_did_change_screen`
TRACE [winit::platform_impl::platform::window_delegate] Completed `windowDidChangeScreen:`
TRACE [winit::platform_impl::platform::window_delegate] Triggered `windowShouldClose:`
TRACE [winit::platform_impl::platform::window_delegate] Completed `windowShouldClose:`
WindowEvent::CloseRequested
TRACE [winit::platform_impl::platform::window] Locked shared state in `fullscreen`
TRACE [winit::platform_impl::platform::window] Unlocked shared state in `fullscreen`
TRACE [winit::platform_impl::platform::window_delegate] Triggered `windowWillClose:`
TRACE [winit::platform_impl::platform::window_delegate] Completed `windowWillClose:`
Event::LoopDestroyed.
-----------------
Creating window "Second Window"
TRACE [winit::platform_impl::platform::window] Triggered `WinitWindow::new`
TRACE [winit::platform_impl::platform::view] Triggered `validAttributesForMarkedText`
TRACE [winit::platform_impl::platform::view] Completed `validAttributesForMarkedText`
TRACE [winit::platform_impl::platform::view] Triggered `validAttributesForMarkedText`
TRACE [winit::platform_impl::platform::view] Completed `validAttributesForMarkedText`
TRACE [winit::platform_impl::platform::view] Triggered `validAttributesForMarkedText`
TRACE [winit::platform_impl::platform::view] Completed `validAttributesForMarkedText`
TRACE [winit::platform_impl::platform::view] Triggered `frameDidChange:`
TRACE [winit::platform_impl::platform::view] Completed `frameDidChange:`
TRACE [winit::platform_impl::platform::view] Triggered `viewDidMoveToWindow`
TRACE [winit::platform_impl::platform::view] Completed `viewDidMoveToWindow`
TRACE [winit::platform_impl::platform::window] Locked shared state in `WinitWindow::new`
TRACE [winit::platform_impl::platform::window] Unlocked shared state in `WinitWindow::new`
TRACE [winit::platform_impl::platform::window] Locked shared state in `set_fullscreen`
TRACE [winit::platform_impl::platform::window] Unlocked shared state in `set_fullscreen`
TRACE [winit::platform_impl::platform::window] Triggered `canBecomeMainWindow`
TRACE [winit::platform_impl::platform::window] Completed `canBecomeMainWindow`
TRACE [winit::platform_impl::platform::window] Triggered `canBecomeKeyWindow`
TRACE [winit::platform_impl::platform::window] Completed `canBecomeKeyWindow`
TRACE [winit::platform_impl::platform::window_delegate] Triggered `windowDidBecomeKey:`
TRACE [winit::platform_impl::platform::window_delegate] Completed `windowDidBecomeKey:`
TRACE [winit::platform_impl::platform::view] Triggered `acceptsFirstResponder`
TRACE [winit::platform_impl::platform::view] Completed `acceptsFirstResponder`
TRACE [winit::platform_impl::platform::view] Triggered `acceptsFirstResponder`
TRACE [winit::platform_impl::platform::view] Completed `acceptsFirstResponder`
TRACE [winit::platform_impl::platform::window] Triggered `canBecomeKeyWindow`
TRACE [winit::platform_impl::platform::window] Completed `canBecomeKeyWindow`
TRACE [winit::platform_impl::platform::window] Triggered `canBecomeKeyWindow`
TRACE [winit::platform_impl::platform::window] Completed `canBecomeKeyWindow`
TRACE [winit::platform_impl::platform::window] Completed `WinitWindow::new`
thread 'main' panicked at 'The panic info must exist here. This failure indicates a developer error.', src/platform_impl/macos/app_state.rs:374:14
stack backtrace:
   0: rust_begin_unwind
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/panicking.rs:64:14
   2: core::panicking::panic_display
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/panicking.rs:135:5
   3: core::panicking::panic_str
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/panicking.rs:119:5
   4: core::option::expect_failed
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/option.rs:1879:5
   5: core::option::Option<T>::expect
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/option.rs:741:21
   6: winit::platform_impl::platform::app_state::AppState::cleared
             at ./src/platform_impl/macos/app_state.rs:372:26
   7: winit::platform_impl::platform::observer::control_flow_end_handler::{{closure}}
             at ./src/platform_impl/macos/observer.rs:79:21
   8: winit::platform_impl::platform::observer::control_flow_handler::{{closure}}
             at ./src/platform_impl/macos/observer.rs:41:9
   9: std::panicking::try::do_call
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:483:40
  10: ___rust_try
  11: std::panicking::try
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:447:19
  12: std::panic::catch_unwind
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panic.rs:137:14
  13: winit::platform_impl::platform::event_loop::stop_app_on_panic
             at ./src/platform_impl/macos/event_loop.rs:245:11
  14: winit::platform_impl::platform::observer::control_flow_handler
             at ./src/platform_impl/macos/observer.rs:39:5
  15: winit::platform_impl::platform::observer::control_flow_end_handler
             at ./src/platform_impl/macos/observer.rs:74:9
  16: <unknown>
  17: <unknown>
  18: <unknown>
  19: <unknown>
  20: <unknown>
  21: <unknown>
  22: <unknown>
  23: <unknown>
  24: <unknown>
  25: <unknown>
  26: <() as objc2::message::MessageArguments>::__invoke
             at /Users/emilk/.cargo/registry/src/github.com-1ecc6299db9ec823/objc2-0.3.0-beta.3/src/message/mod.rs:390:26
  27: objc2::message::platform::send_unverified::{{closure}}
             at /Users/emilk/.cargo/registry/src/github.com-1ecc6299db9ec823/objc2-0.3.0-beta.3/src/message/apple/mod.rs:36:33
  28: objc2::message::conditional_try
             at /Users/emilk/.cargo/registry/src/github.com-1ecc6299db9ec823/objc2-0.3.0-beta.3/src/message/mod.rs:30:5
  29: objc2::message::platform::send_unverified
             at /Users/emilk/.cargo/registry/src/github.com-1ecc6299db9ec823/objc2-0.3.0-beta.3/src/message/apple/mod.rs:36:14
  30: objc2::message::MessageReceiver::send_message
             at /Users/emilk/.cargo/registry/src/github.com-1ecc6299db9ec823/objc2-0.3.0-beta.3/src/message/mod.rs:195:46
  31: winit::platform_impl::platform::appkit::application::NSApplication::run
             at /Users/emilk/.cargo/registry/src/github.com-1ecc6299db9ec823/objc2-0.3.0-beta.3/src/macros/__rewrite_self_arg.rs:42:21
  32: winit::platform_impl::platform::event_loop::EventLoop<T>::run_return::{{closure}}
             at ./src/platform_impl/macos/event_loop.rs:220:22
  33: objc2::rc::autorelease::autoreleasepool
             at /Users/emilk/.cargo/registry/src/github.com-1ecc6299db9ec823/objc2-0.3.0-beta.3/src/rc/autorelease.rs:313:5
  34: winit::platform_impl::platform::event_loop::EventLoop<T>::run_return
             at ./src/platform_impl/macos/event_loop.rs:211:25
  35: <winit::event_loop::EventLoop<T> as winit::platform::run_return::EventLoopExtRunReturn>::run_return
             at ./src/platform/run_return.rs:51:9
  36: emilk::main::show_window
             at ./examples/emilk.rs:23:9
  37: emilk::main
             at ./examples/emilk.rs:71:5
  38: core::ops::function::FnOnce::call_once
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/ops/function.rs:507:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/platform_impl/macos/event_loop.rs:253:55
stack backtrace:
   0: rust_begin_unwind
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/panicking.rs:64:14
   2: core::panicking::panic
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/panicking.rs:111:5
   3: core::option::Option<T>::unwrap
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/option.rs:778:21
   4: winit::platform_impl::platform::event_loop::stop_app_on_panic
             at ./src/platform_impl/macos/event_loop.rs:253:34
   5: winit::platform_impl::platform::observer::control_flow_handler
             at ./src/platform_impl/macos/observer.rs:39:5
   6: winit::platform_impl::platform::observer::control_flow_end_handler
             at ./src/platform_impl/macos/observer.rs:74:9
   7: <unknown>
   8: <unknown>
   9: <unknown>
  10: <unknown>
  11: <unknown>
  12: <unknown>
  13: <unknown>
  14: <unknown>
  15: <unknown>
  16: <unknown>
  17: <() as objc2::message::MessageArguments>::__invoke
             at /Users/emilk/.cargo/registry/src/github.com-1ecc6299db9ec823/objc2-0.3.0-beta.3/src/message/mod.rs:390:26
  18: objc2::message::platform::send_unverified::{{closure}}
             at /Users/emilk/.cargo/registry/src/github.com-1ecc6299db9ec823/objc2-0.3.0-beta.3/src/message/apple/mod.rs:36:33
  19: objc2::message::conditional_try
             at /Users/emilk/.cargo/registry/src/github.com-1ecc6299db9ec823/objc2-0.3.0-beta.3/src/message/mod.rs:30:5
  20: objc2::message::platform::send_unverified
             at /Users/emilk/.cargo/registry/src/github.com-1ecc6299db9ec823/objc2-0.3.0-beta.3/src/message/apple/mod.rs:36:14
  21: objc2::message::MessageReceiver::send_message
             at /Users/emilk/.cargo/registry/src/github.com-1ecc6299db9ec823/objc2-0.3.0-beta.3/src/message/mod.rs:195:46
  22: winit::platform_impl::platform::appkit::application::NSApplication::run
             at /Users/emilk/.cargo/registry/src/github.com-1ecc6299db9ec823/objc2-0.3.0-beta.3/src/macros/__rewrite_self_arg.rs:42:21
  23: winit::platform_impl::platform::event_loop::EventLoop<T>::run_return::{{closure}}
             at ./src/platform_impl/macos/event_loop.rs:220:22
  24: objc2::rc::autorelease::autoreleasepool
             at /Users/emilk/.cargo/registry/src/github.com-1ecc6299db9ec823/objc2-0.3.0-beta.3/src/rc/autorelease.rs:313:5
  25: winit::platform_impl::platform::event_loop::EventLoop<T>::run_return
             at ./src/platform_impl/macos/event_loop.rs:211:25
  26: <winit::event_loop::EventLoop<T> as winit::platform::run_return::EventLoopExtRunReturn>::run_return
             at ./src/platform/run_return.rs:51:9
  27: emilk::main::show_window
             at ./examples/emilk.rs:23:9
  28: emilk::main
             at ./examples/emilk.rs:71:5
  29: core::ops::function::FnOnce::call_once
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/ops/function.rs:507:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
TRACE [winit::platform_impl::platform::window] Locked shared state in `fullscreen`
TRACE [winit::platform_impl::platform::window] Unlocked shared state in `fullscreen`
TRACE [winit::platform_impl::platform::window_delegate] Triggered `windowWillClose:`
TRACE [winit::platform_impl::platform::window_delegate] Completed `windowWillClose:`
```

**Comment by rib** [comment]
Just a stab in the dark atm but looking at the macos backend it has a double buffered control flow status (i.e the wait or exit status) for the event loop. (https://github.com/rust-windowing/winit/blob/b18295a1ce1efbf1d0ab6551b5df1d7e6c511d50/src/platform_impl/macos/app_state.rs#L119)

I suspicious atm about how that's being updated and read.

Setting the status to exit is what will cause the loop to exit after the first window is closed but then when re-running the loop it's notable that the status is only changed once the first event is read and dispatched.

As far as winit is concerned I think there's chance that the control flow status might be effectively read as "exit" when re-running the loop, before there is an opportunity to reset it.

Just a guess / hunch at the moment but I don't see anything that resets the control flow state in `run_return`. Initially it is simply `Default` initialized and stored in a global variable called `HANDLER` app_state.rs. I guess maybe `run_return` needs to make sure to reset this via some added `AppState` API

**Comment by rib** [comment]
there are two callbacks for the second window though before it exits the loop so maybe that's a red herring.
